File: crates/deepsec-cli/src/commands/data_commit.rs
use crate::config_loader::Context;
use anyhow::{Context as _, Result, anyhow};
use clap::Args as ClapArgs;
use colored::Colorize;
use std::process::Command;

#[derive(Debug, ClapArgs)]
pub struct Args {
    /// Subset of the data dir to commit. Defaults to the entire
    /// data dir.
    #[arg(long)]
    pub paths: Option<Vec<String>>,
    /// Commit message. Defaults to a generated one.
    #[arg(short, long)]
    pub message: Option<String>,
    /// Repo to operate on. Defaults to the directory containing
    /// the config file (or cwd).
    #[arg(long)]
    pub repo: Option<std::path::PathBuf>,
    /// Don't fail if there is nothing to commit.
    #[arg(long)]
    pub allow_empty: bool,
}

pub fn run(args: Args, ctx: &Context) -> Result<()> {
    let repo = args.repo.unwrap_or_else(|| {
        ctx.config_path
            .as_ref()
            .and_then(|p| p.parent().map(|p| p.to_path_buf()))
            .unwrap_or_else(|| ctx.cwd.clone())
    });

    if !repo.join(".git").exists() {
        return Err(anyhow!(
            "{} is not a git repository (pass --repo)",
            repo.display()
        ));
    }

    let data_path = ctx.data_root.as_path();
    let rel_data = data_path
        .strip_prefix(&repo)
        .map(|p| p.to_path_buf())
        .unwrap_or_else(|_| data_path.to_path_buf());

    let mut add = Command::new("git");
    add.arg("-C").arg(&repo).arg("add").arg("--");
    if let Some(paths) = args.paths {
        for p in paths {
            add.arg(p);
        }
    } else {
        add.arg(&rel_data);
    }
    let status = add.status().context("running git add")?;
    if !status.success() {
        return Err(anyhow!("git add failed"));
    }

    // Check if there's anything staged.
    let diff = Command::new("git")
        .arg("-C")
        .arg(&repo)
        .arg("diff")
        .arg("--cached")
        .arg("--quiet")
        .status()
        .context("running git diff --cached")?;
    if diff.success() {
        // exit 0 → nothing staged
        if args.allow_empty {
            println!("{} no changes to commit", "data-commit".dimmed());
            return Ok(());
        }
        return Err(anyhow!("no changes staged"));
    }

    let message = args.message.unwrap_or_else(|| {
        format!("deepsec: update {} ({})",
            rel_data.display(),
            time::OffsetDateTime::now_utc()
                .format(&time::format_description::well_known::Iso8601::DEFAULT)
                .unwrap_or_default())
    });

    let commit = Command::new("git")
        .arg("-C")
        .arg(&repo)
        .arg("commit")
        .arg("-m")
        .arg(&message)
        .status()
        .context("running git commit")?;
    if !commit.success() {
        return Err(anyhow!("git commit failed"));
    }
    println!("{} {}", "data-commit".bold().green(), message);
    Ok(())
}

File: internal/cli/commands/data_commit.go
package commands

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"time"

	"github.com/noeljackson/deepsec/internal/cli"
	"github.com/spf13/cobra"
)

// NewDataCommitCmd does `git add data/ && git commit`.
func NewDataCommitCmd(loader func() (*cli.Context, error)) *cobra.Command {
	var message, repo string
	var paths []string
	var allowEmpty bool
	cmd := &cobra.Command{
		Use:   "data-commit",
		Short: "git add data/ + git commit",
		RunE: func(cmd *cobra.Command, _ []string) error {
			ctx, err := loader()
			if err != nil {
				return err
			}
			repoPath := repo
			if repoPath == "" {
				if ctx.ConfigPath != "" {
					repoPath = filepath.Dir(ctx.ConfigPath)
				} else {
					repoPath = ctx.CWD
				}
			}
			if _, err := os.Stat(filepath.Join(repoPath, ".git")); err != nil {
				return fmt.Errorf("%s is not a git repository", repoPath)
			}

			addArgs := []string{"-C", repoPath, "add", "--"}
			if len(paths) > 0 {
				addArgs = append(addArgs, paths...)
			} else {
				relData, err := filepath.Rel(repoPath, ctx.DataRoot.Path)
				if err != nil {
					relData = ctx.DataRoot.Path
				}
				addArgs = append(addArgs, relData)
			}
			if err := exec.Command("git", addArgs...).Run(); err != nil {
				return fmt.Errorf("git add: %w", err)
			}

			diff := exec.Command("git", "-C", repoPath, "diff", "--cached", "--quiet")
			if err := diff.Run(); err == nil {
				if allowEmpty {
					fmt.Println("data-commit: nothing to commit")
					return nil
				}
				return fmt.Errorf("no changes staged")
			}

			if message == "" {
				message = fmt.Sprintf("deepsec: update %s (%s)",
					ctx.DataRoot.Path,
					time.Now().UTC().Format(time.RFC3339))
			}
			commit := exec.Command("git", "-C", repoPath, "commit", "-m", message)
			commit.Stdout = os.Stdout
			commit.Stderr = os.Stderr
			if err := commit.Run(); err != nil {
				return fmt.Errorf("git commit: %w", err)
			}
			fmt.Printf("data-commit %s\n", message)
			return nil
		},
	}
	cmd.Flags().StringVarP(&message, "message", "m", "", "Commit message")
	cmd.Flags().StringVar(&repo, "repo", "", "Repo path (default: config dir or cwd)")
	cmd.Flags().StringSliceVar(&paths, "paths", nil, "Paths to add (default: the entire data dir)")
	cmd.Flags().BoolVar(&allowEmpty, "allow-empty", false, "Exit cleanly when there's nothing to commit")
	return cmd
}
